S3commander allow sts rev1

djkiourtsis opened the pull request:
Add a component to allow the use of sts credentials with s3commander.

djkiourtsis committed:
Add optional session token to bucket component

This allows the use of aws sessions with s3commander.  If sessions are
not being used, then the session token defaults to null.

djkiourtsis committed:
Update s3commander version to 0.13.0

This version adds support for aws sessions.

djkiourtsis committed:
Regenerate AWS backend on credentials change

If the AWS credential bindings change while s3commander is running,
then we want to collect the changes and update the AWS backend.  This
will maily be used to allow us to refresh temporary credentials.

djkiourtsis committed:
Add component for use with sts credentials

This adds an STS session component to s3commander.  This component
is similar to the bucket component, but will take in an api url and
optional header.  This url must return the AWS credentials for an
sts session, including the session token and duration.  The component
itself will render a bucket with the temporary credentials.

djkiourtsis committed:
Add time based refresh for sts credentials

Sts credentials will expire after an hour by default.  To account for
this, we want to attempt to collect a new set of sts credentials
befor the existing credentials expire.

djkiourtsis commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add STS support to s3commander](https://trello.com/c/OCeMYDOU/24-add-sts-support-to-s3commander)

CtrlC-Root commented on src/stssession/STSSessionController.ts:
```diff
@@ -0,0 +1,141 @@
+
+    // Set http headers
+    var httpHeaders;
+    if (this.stsHeaderName === undefined || this.stsHeaderValue === undefined) {
```
Create the default headers up front:
`headers = {'Content-Type': 'application/json'}`

then add to them if you need to:
`headers[this.stsHeaderName] = this.stsHeaderValue`

CtrlC-Root commented on src/stssession/STSSessionController.ts:
```diff
@@ -0,0 +1,141 @@
+      this.stsApiUrl,
+      {},
+      httpHeaders)
+    .then((response: any): ng.IPromise<any> => this.refreshSTSCredentials(response));
```
What if it fails?

CtrlC-Root replied:
Also as far as I can tell this is only used in one place so you could just inline the function as an anonymous function.

CtrlC-Root replied:
Also don't use `any`. You can write an interface and use it here or don't return anything at all from the function.

CtrlC-Root commented on src/stssession/STSSessionController.ts:
```diff
@@ -0,0 +1,141 @@
+      };
+    }
+
+    var result: ng.IPromise<any> = this.httpService.post(
```
Don't use `any` unless you have to. You can look up the type to expect here.

CtrlC-Root commented on src/stssession/STSSessionController.ts:
```diff
@@ -103,28 +118,28 @@ export class STSSessionController {
+    if (!this.stsHeaderName === undefined && !this.stsHeaderValue === undefined) {
       var name = this.stsHeaderName;
-      httpHeaders = {
+      this.httpHeaders = {
```
I made an earlier comment on this but I see you're changing this so you can ignore the previous one.

CtrlC-Root reviewed (changes requested):
I commented on some minor improvements.

djkiourtsis committed:
Add stssession error handling

This will handle and display any http errors that may occur while
attempting to set up the sts session.

mliudev commented on src/stssession/STSSessionController.ts:
```diff
@@ -86,6 +96,16 @@ export class STSSessionController {
+    };
+
+    // collect credentials from api every 60 minutes
+    setInterval(() => this.credentialsFromAPI(), 1000 * 60 * 60);
```
Should probably refresh before the credentials in order to give some buffer for getting and refreshing the token.

djkiourtsis replied:
This is a good idea; I'll get an early refresh in the code.

mliudev commented on src/stssession/STSSessionController.ts:
```diff
@@ -103,39 +123,26 @@ export class STSSessionController {
+      this.stsApiUrl,
+      {},
+      this.httpHeaders)
+    .then((response: any): void => {
```
According to @CtrlC-Root's previous reviews you shouldn't use 'any' here since you should know what type to expect.

djkiourtsis replied:
We may not have a choice with this use of any.  `response` is an http response, and our current version of angular (1.X) does not have enhanced http handling.

mliudev replied:
I guess it's not easy trying to figure out what the type of the response is for AngularJS since there isn't exactly a ts file that it defines. I'm OK with this then @CtrlC-Root thoughts?

CtrlC-Root replied:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/angular/index.d.ts#L1503

Could the definition be in here somewhere?

djkiourtsis replied:
Found the definition from there; thanks!

mliudev commented on src/stssession/STSSessionController.ts:
```diff
@@ -143,6 +143,10 @@ export class STSSessionController {
       this.awsSecretAccessKey = response.data.SecretAccessKey;
       this.awsSessionToken = response.data.SessionToken;
+    })
+    .catch((response: any) => {
```
Don't use any.

mliudev reviewed (commented):
Left some additional comments in addition to @CtrlC-Root 's.

djkiourtsis committed:
Shorten the STS credential refresh delay to 55 minutes

By default, STS sessions will expire after an hour.  This will shorten
the STS credential refresh delay so that we can ensure a new set of
credentials before the existing credentials expire.

djkiourtsis committed:
Prevent STS session from prematurely rendering s3commander

This prevents the STS session from rendering the s3commander bucket
before the credentials are gathered.  Prematurely rendering the bucket
results in a race condition that may cause the bucket to error.
